worker-testing-util/src/main/java/com/hpe/caf/worker/testing/WorkerServicesFactory.java

```diff
@@ -46,7 +47,8 @@ public static WorkerServices create() throws ModuleLoaderException, CipherExcept
         Cipher cipher = ModuleLoader.getService(CipherProvider.class, NullCipherProvider.class).getCipher(bootstrapConfiguration);
         ServicePath path = bootstrapConfiguration.getServicePath();
         ConfigurationSource configurationSource = ModuleLoader.getService(ConfigurationSourceProvider.class).getConfigurationSource(bootstrapConfiguration, cipher, path, decoder);
-        DataStore dataStore = ModuleLoader.getService(DataStoreProvider.class).getDataStore(configurationSource);
+//        DataStore dataStore = ModuleLoader.getService(DataStoreProvider.class).getDataStore(configurationSource);
+        DataStore dataStore = new HttpDataStoreProvider().getDataStore(configurationSource);
```

Comment:
TODO: Remove, temporarily added to try and test

Reply:
We may need to use one of the older workers (TextExtract, OCR, Speech) after all.

Reply:
Thanks, I updated worker-textextract to pull in these changes in a temp branch:

https://github.houston.softwaregrp.net/caf/worker-textextract/compare/SCMOD-13868-temp

Comparing the logs for one of the integration tests on master:

http://sou-jenkins2.swinfra.net/job/caf/job/caf~worker-textextract~develop~CI/833/consoleFull

```
===============  Starting test Item bcc.eml ======================
13:19:58.440 webdav> 172.17.0.7:80 16.73.32.72 - - [21/May/2021:17:19:58 +0000] "PROPFIND /webdav/c9266a40-ce23-4dc3-b378-15321a0481e6 HTTP/1.1" 404 536 "-" "Microsoft-WebDAV-MiniRedir/6.1.7601"
13:19:58.451 webdav> 172.17.0.7:80 16.73.32.72 - - [21/May/2021:17:19:58 +0000] "PROPFIND /webdav HTTP/1.1" 301 589 "-" "Microsoft-WebDAV-MiniRedir/6.1.7601"
13:19:58.453 webdav> 172.17.0.7:80 16.73.32.72 - - [21/May/2021:17:19:58 +0000] "PROPFIND /webdav/ HTTP/1.1" 207 1052 "-" "Microsoft-WebDAV-MiniRedir/6.1.7601"
13:19:58.471 webdav> 172.17.0.7:80 16.73.32.72 - - [21/May/2021:17:19:58 +0000] "PUT /webdav/c9266a40-ce23-4dc3-b378-15321a0481e6 HTTP/1.1" 201 364 "-" "Microsoft-WebDAV-MiniRedir/6.1.7601"
13:19:58.486 webdav> 172.17.0.7:80 16.73.32.72 - - [21/May/2021:17:19:58 +0000] "PROPPATCH /webdav/c9266a40-ce23-4dc3-b378-15321a0481e6 HTTP/1.1" 207 641 "-" "Microsoft-WebDAV-MiniRedir/6.1.7601"
13:19:58.510 webdav> 172.17.0.7:80 16.73.32.72 - - [21/May/2021:17:19:58 +0000] "HEAD /webdav/c9266a40-ce23-4dc3-b378-15321a0481e6 HTTP/1.1" 200 237 "-" "Microsoft-WebDAV-MiniRedir/6.1.7601"
13:19:58.531 webdav> 172.17.0.7:80 16.73.32.72 - - [21/May/2021:17:19:58 +0000] "PUT /webdav/c9266a40-ce23-4dc3-b378-15321a0481e6 HTTP/1.1" 204 150 "-" "Microsoft-WebDAV-MiniRedir/6.1.7601"
13:19:58.550 webdav> 172.17.0.7:80 16.73.32.72 - - [21/May/2021:17:19:58 +0000] "PROPPATCH /webdav/c9266a40-ce23-4dc3-b378-15321a0481e6 HTTP/1.1" 207 641 "-" "Microsoft-WebDAV-MiniRedir/6.1.7601"
================================================================================
 QUEUEING NEW TASK: bcc.eml
================================================================================
13:19:58.564 worker-textextract> INFO  [2021-05-21 17:19:58,563] com.hpe.caf.worker.textextract.WorkerTextExtract: Beinging to process file...
13:19:58.593 worker-textextract> INFO  [2021-05-21 17:19:58,589] com.hpe.caf.worker.textextract.process.OperationExecutor: About to perform ExtractFileType operation.
13:19:58.743 worker-textextract> INFO  [2021-05-21 17:19:58,742] com.hpe.caf.worker.textextract.process.ExtractFileTypeUsingKeyView: File type detected: message/rfc822
13:19:58.743 worker-textextract> INFO  [2021-05-21 17:19:58,742] com.hpe.caf.worker.textextract.process.OperationExecutor: About to perform ExtractMetadata operation.
13:19:58.753 worker-textextract> INFO  [2021-05-21 17:19:58,753] com.hpe.caf.worker.textextract.process.OperationExecutor: About to perform ExtractText operation.
13:19:58.754 worker-textextract> INFO  [2021-05-21 17:19:58,753] com.hpe.caf.worker.textextract.process.ExtractText: About to perform text extraction.
13:19:58.759 worker-textextract> WARN  [2021-05-21 17:19:58,757] com.hpe.caf.worker.textextract.childextraction.KeyviewSpecialCaseChildExtractor: Skipping subfile: root/
13:19:59.502 worker-textextract> INFO  [2021-05-21 17:19:59,502] com.hpe.caf.worker.textextract.process.ExtractText: Text extraction complete.
13:19:59.502 worker-textextract> INFO  [2021-05-21 17:19:59,502] com.hpe.caf.worker.textextract.process.OperationExecutor: About to perform ExtractChildren operation.
13:19:59.502 worker-textextract> INFO  [2021-05-21 17:19:59,502] com.hpe.caf.worker.textextract.process.ExtractChildren: About to perform child extraction
13:19:59.503 worker-textextract> WARN  [2021-05-21 17:19:59,503] com.hpe.caf.worker.textextract.childextraction.KeyviewSpecialCaseChildExtractor: Skipping subfile: root/
13:19:59.508 worker-textextract> INFO  [2021-05-21 17:19:59,507] com.hpe.caf.worker.textextract.process.ExtractChildren: Child extraction complete
13:19:59.508 worker-textextract> INFO  [2021-05-21 17:19:59,507] com.hpe.caf.worker.textextract.WorkerTextExtract: Processing of file complete.
New deliverybcc.eml, status: RESULT_SUCCESS
New delivery: task id: bcc.eml, status: RESULT_SUCCESS
*** Validating 'fieldChanges'
*** Created validator: PropertyMapValidator
*** Validating 'CHILD_INFO_0_BINARY'
*** Created validator: DocumentWorkerFieldChangesValidator
About to retrieve content for com.hpe.caf.util.ref.ReferencedData@40b48477
13:19:59.863 [Thread-3] DEBUG com.hpe.caf.worker.datastore.fs.FileSystemDataStore - Requesting TEST_CONTAINER/cd144184-807c-438f-83c4-cffb5061e464
13:19:59.602 webdav> 172.17.0.7:80 16.73.32.72 - - [21/May/2021:17:19:59 +0000] "PROPFIND /webdav/TEST_CONTAINER/cd144184-807c-438f-83c4-cffb5061e464 HTTP/1.1" 207 1097 "-" "Microsoft-WebDAV-MiniRedir/6.1.7601"
13:19:59.623 webdav> 172.17.0.7:80 16.73.32.72 - - [21/May/2021:17:19:59 +0000] "GET /webdav/TEST_CONTAINER/cd144184-807c-438f-83c4-cffb5061e464 HTTP/1.1" 200 8116 "-" "Microsoft-WebDAV-MiniRedir/6.1.7601"
Finished retrieving content for com.hpe.caf.util.ref.ReferencedData@40b48477
```

with the logs from the temp branch:

http://sou-jenkins2.swinfra.net/job/caf/view/Developer/job/caf~worker-textextract~SCMOD-13868-temp~CI/1/consoleFull

```
===============  Starting test Item bcc.eml ======================
06:20:03.653 [main] DEBUG com.hpe.caf.worker.datastore.http.HttpDataStore - Received store request for partial reference (may be null, which is allowed): null
06:20:03.669 [main] DEBUG com.hpe.caf.worker.datastore.http.HttpDataStore - Successfully stored data with reference: a68d6f1d-65ce-4161-83e3-dbdf56961c47
06:20:03.661 webdav> 172.17.0.6:80 16.73.32.63 - - [08/Jun/2021:10:20:03 +0000] "PUT /webdav/a68d6f1d-65ce-4161-83e3-dbdf56961c47 HTTP/1.1" 201 365 "-" "Java/1.8.0_111-3-ojdkbuild"
================================================================================
 QUEUEING NEW TASK: bcc.eml
================================================================================
06:20:03.666 worker-textextract> INFO  [2021-06-08 10:20:03,666] com.hpe.caf.worker.textextract.WorkerTextExtract: Beinging to process file...
06:20:03.689 worker-textextract> INFO  [2021-06-08 10:20:03,689] com.hpe.caf.worker.textextract.process.OperationExecutor: About to perform ExtractFileType operation.
06:20:03.782 worker-textextract> INFO  [2021-06-08 10:20:03,782] com.hpe.caf.worker.textextract.process.ExtractFileTypeUsingKeyView: File type detected: message/rfc822
06:20:03.782 worker-textextract> INFO  [2021-06-08 10:20:03,782] com.hpe.caf.worker.textextract.process.OperationExecutor: About to perform ExtractMetadata operation.
06:20:03.791 worker-textextract> INFO  [2021-06-08 10:20:03,790] com.hpe.caf.worker.textextract.process.OperationExecutor: About to perform ExtractText operation.
06:20:03.791 worker-textextract> INFO  [2021-06-08 10:20:03,790] com.hpe.caf.worker.textextract.process.ExtractText: About to perform text extraction.
06:20:03.791 worker-textextract> WARN  [2021-06-08 10:20:03,791] com.hpe.caf.worker.textextract.childextraction.KeyviewSpecialCaseChildExtractor: Skipping subfile: root/
06:20:04.358 worker-textextract> INFO  [2021-06-08 10:20:04,357] com.hpe.caf.worker.textextract.process.ExtractText: Text extraction complete.
06:20:04.358 worker-textextract> INFO  [2021-06-08 10:20:04,357] com.hpe.caf.worker.textextract.process.OperationExecutor: About to perform ExtractChildren operation.
06:20:04.358 worker-textextract> INFO  [2021-06-08 10:20:04,357] com.hpe.caf.worker.textextract.process.ExtractChildren: About to perform child extraction
06:20:04.358 worker-textextract> WARN  [2021-06-08 10:20:04,358] com.hpe.caf.worker.textextract.childextraction.KeyviewSpecialCaseChildExtractor: Skipping subfile: root/
06:20:04.366 worker-textextract> INFO  [2021-06-08 10:20:04,366] com.hpe.caf.worker.textextract.process.ExtractChildren: Child extraction complete
06:20:04.367 worker-textextract> INFO  [2021-06-08 10:20:04,366] com.hpe.caf.worker.textextract.WorkerTextExtract: Processing of file complete.
New deliverybcc.eml, status: RESULT_SUCCESS
New delivery: task id: bcc.eml, status: RESULT_SUCCESS
*** Validating 'fieldChanges'
*** Created validator: PropertyMapValidator
*** Validating 'CHILD_INFO_0_BINARY'
*** Created validator: DocumentWorkerFieldChangesValidator
About to retrieve content for com.hpe.caf.util.ref.ReferencedData@daa2d67
06:20:04.450 [Thread-3] DEBUG com.hpe.caf.worker.datastore.http.HttpDataStore - Received retrieve request for data with reference: TEST_CONTAINER/1e1be598-5a2c-44fe-aa9c-6a2fb320dd75
06:20:04.455 webdav> 172.17.0.6:80 16.73.32.63 - - [08/Jun/2021:10:20:04 +0000] "GET /webdav/TEST_CONTAINER/1e1be598-5a2c-44fe-aa9c-6a2fb320dd75 HTTP/1.1" 200 8117 "-" "Java/1.8.0_111-3-ojdkbuild"
06:20:04.466 [Thread-3] DEBUG com.hpe.caf.worker.datastore.http.HttpDataStore - Successfully retrieved data with reference: TEST_CONTAINER/1e1be598-5a2c-44fe-aa9c-6a2fb320dd75
Finished retrieving content for com.hpe.caf.util.ref.ReferencedData@daa2d67
```

I can see that the integration tests are using the new HttpDataStore class successfully, performing GETs and PUTs with it, so am fairly happy this is working as expected.

Image showing above logs side by side, develop branch on left, new temp branch on right:

![image](https://user-images.githubusercontent.com/55391479/121174774-30691a80-c852-11eb-83b7-f9c5fc1b057f.png)

One other question I had: 

The integration tests call `WorkerServicesFactory` via this method in their setup:

https://github.houston.softwaregrp.net/caf/worker-textextract/blob/develop/worker-textextract-container/src/test/java/com/hpe/caf/worker/textextract/WorkerTextExtractAcceptanceIT.java#L26

```
controller = TestRunnerSingle.getTestController(testControllerProvider, new SystemSettingsProvider().getBooleanSetting("testCaseGeneration", false));
```

The `WorkerServicesFactory` is the class the integration tests use to get an instance of a `DataStore`.

I couldn't find a way to make the integration tests use `worker-store-http`, while still making the worker use `worker-store-fs` without hardcoding the `HttpDataStore` class here in `WorkerServices`?

Reply:
> I couldn't find a way to make the integration tests use `worker-store-http`, while still making the worker use `worker-store-fs` without hardcoding the `HttpDataStore` class here in `WorkerServicesFactory`?

Hard-coding it is technically a breaking change.  You could introduce a `worker-testing` config setting that if set uses the `HttpDataStore` and if not set defaults to the current behaviour.

Reply:
Thanks, when you say config setting do you mean a system property, or am I misunderstanding? e.g:

In `WorkerServicesFactory` in worker-framework:
```
DataStore dataStore = new SystemSettingsProvider().getBooleanSetting("worker.testing") 
            ? new HttpDataStoreProvider().getDataStore(configurationSource)
            :  ModuleLoader.getService(DataStoreProvider.class).getDataStore(configurationSource);
```

Then in the worker's container pom.xml:

```
<!-- Properties for the worker. -->
    <properties>
        <worker.testing>true</worker.testing>
        ...
    </properties>
```

Reply:
A system property is fine.  But you wouldn't call it `worker.testing`.  It would be more like `worker.testing.usehttpdatastore`.  Or `worker.testing.datastore` with a value of `http`.

Reply:
Thanks, updated to use a system property now.